### PR TITLE
build: compile only the HAL files needed by the board

### PR DESCRIPTION
**Pull request by cmarcelo**
Instead of requiring every program to specify the HAL related modules,
let the build system do it (in practice everything we compiled depended
on HAL anyway). This allow including only the necessary files in the
compilation.

### CODE REVIEW

**Comment by cmarcelo** [comment]
I haven't splitted in multiple commits because to keep the intermediate stages compiling would be necessary to make further changes (filtering repeated entries) that are completely unrelated to the goal.

I think for this kind of change (a infrastructure change that affects all modules) should be OK to not split per module directory.

**Comment by cmarcelo** [comment]
One side-effect is that the builds will be slightly faster because less files will have to be opened and checked.

Most `#ifdef` in the HAL implementations could go away after this change.

File: mk/sketch_sources.mk [hunk 1]
```diff
@@ -21,6 +21,48 @@ SKETCHOBJS := $(addsuffix .o,$(basename $(SKETCHOBJS)))
...
+
+ifeq ($(HAL_BOARD),HAL_BOARD_APM1)
+LIBTOKENS += \
+	AP_HAL_APM
+endif
+
+ifeq ($(HAL_BOARD),HAL_BOARD_APM2)
+LIBTOKENS += \
+	AP_HAL_APM
+endif
```
Comment (lucasdemarchi) [inline mk/sketch_sources.mk]: I don't expect this kind of change to be backported so maybe we should just not add APM1 and APM2 here?

**Comment by lucasdemarchi** [comment]
Besides my comment  above (which should be up to discussion) I like it very much! Thanks.

**Comment by lucasdemarchi** [comment]
@tridge I'm assigning this one to you. It fixes several places in which the libraries in examples and elsewhere were not kept up-to-date.

Regarding the split commits, can we talk about that soon? This kind of change IMO deserves a waiver because splitting the commits here do more harm than help (it's certainly not easier to review 40+ commits doing the exact same thing nor it helps preventing conflicts).

**Comment by tridge** [comment]
this was included in pr#3042

**Pull request closed without merging**